File: C++/matrix/sim.cpp
template<typename T> struct sim{
    struct arr{
        int sz;
        T sum;
        vector<T> pref;
        arr(int n) : sz(n), sum(T()), pref(n, T()){}
    };

    int N, block;
    vector<arr> table;
    vector<T> block_pref;

    sim(int n) : N(n), block(sqrt(n)+1){
        int num_blocks = (N+block-1)/block;
        for (int i = 0; i < num_blocks; i++){
            int sz = min(block, N-i*block);
            table.emplace_back(sz);
        }
        block_pref.assign(num_blocks, T());
    }

    void add(int i, T x){
        int y = i/block, z = i%block;
        table[y].sum += x;
        for (int j = z; j < table[y].sz; j++) table[y].pref[j] += x;
        for (int k = y; k < (int)block_pref.size(); k++) block_pref[k] += x;
    }

    T _sum(int r){
        if (r < 0) return T();
        int y = r/block, z = r%block;
        T res = table[y].pref[z];
        if (y > 0) res += block_pref[y-1];
        return res;
    }

    T sum(int l, int r){
        return _sum(r-1) - _sum(l-1);
    }
};

File: py/sim.py
import math

class sim:
    class arr:
        def __init__(self, n):
            self.sz = n
            self.sum = 0
            self.pref = [0]*n

    def __init__(self, n):
        self.N = n
        self.block = int(math.sqrt(n))+1
        self.table = []
        num_blocks = (self.N+self.block-1)//self.block
        self.block_pref = [0]*num_blocks
        for i in range(num_blocks):
            sz = min(self.block, self.N-i*self.block)
            self.table.append(sim.arr(sz))

    def add(self, i, x):
        y, z = i//self.block, i%self.block
        self.table[y].sum += x
        for j in range(z, self.table[y].sz):
            self.table[y].pref[j] += x
        for k in range(y, len(self.block_pref)):
            self.block_pref[k] += x

    def _sum(self, r):
        if r < 0:
            return 0
        y, z = r//self.block, r%self.block
        res = self.table[y].pref[z]
        if y > 0:
            res += self.block_pref[y-1]
        return res
    
    def sum(self, l, r):
        return self._sum(r-1)-self._sum(l-1)
